### ISSUE TITLE
Docker Daemon User Remapping Issue

### ISSUE THREAD

**Issue by gambol99** [moby/moby#36648]
**Description**

When running the docker daemon we `-userns-remap=default` users are unable to remove files from previous layers. Removing the `--userns-remap=default` fixes the issue.

**Steps to reproduce the issue:**
1. Run a docker daemon with user-remapping
```shell
$ docker run --privileged -ti -v /lib/modules:/lib/modules:ro --rm docker:17.07.0-dind dockerd --storage-driver=overlay2 --userns-remap=default
...
INFO[0001] Docker daemon                                 commit=8784753 graphdriver(s)=overlay2 version=17.07.0-ce
INFO[0001] Daemon has completed initialization          
INFO[0001] API listen on /var/run/docker.sock   
```
2. From another terminal
```shell
$ docker ps
CONTAINER ID        IMAGE                 COMMAND                  CREATED             STATUS              PORTS               NAMES
85a57946a7f6        docker:17.07.0-dind   "dockerd-entrypoin..."   49 seconds ago      Up 49 seconds       2375/tcp            keen_jackson
docker exec -ti 85a57946a7f6 sh
/ # mkdir config
/ # cd config
/config # cat > Dockerfile
FROM alpine
RUN touch /me
RUN rm -fv /me
Sending build context to Docker daemon  2.048kB
Step 1/3 : FROM alpine
latest: Pulling from library/alpine
ff3a5c916c92: Pull complete 
Digest: sha256:7b848083f93822dd21b0a2f14a110bd99f6efb4b838d499df6d04a49d0debf8b
Status: Downloaded newer image for alpine:latest
 ---> 3fd9065eaf02
Step 2/3 : RUN touch /me
 ---> Running in a65f9e7f6e31
 ---> ac8dc1d09597
Removing intermediate container a65f9e7f6e31
Step 3/3 : RUN rm -fv /me
 ---> Running in da8381659941
removed '/me'
 ---> 1f1d3637d8fc
Removing intermediate container da8381659941
Successfully built 1f1d3637d8fc
Successfully tagged test:latest
/config # docker run -ti test ls /me -l
-rw-r--r--    1 root     root             0 Mar 20 21:47 /me
```

Note this logged in the docker daemon 
```shell
ERRO[0118] Can't add file /var/lib/docker/165536.165536/overlay2/c4b79d5db0d66be05dd45290b64de80c81351378c2eda0f84aa507b74b8b7317/diff/me to tar: Host ID 0 cannot be mapped to a container ID 
```

**Describe the results you received:**
The file remained, probably due to the later not being saved?

**Describe the results you expected:**
I didn't expect the file to exist

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**
```shell
/config # docker info
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 3
Server Version: 17.07.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 3addd840653146c90a254301d6c3a663c7fd6429
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
 userns
Kernel Version: 4.16.0-0.rc5.git3.1.fc29.x86_64
Operating System: Alpine Linux v3.6 (containerized)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.52GiB
Name: 85a57946a7f6
ID: OP3S:TFFE:BUE6:HPAQ:D24F:VJ5K:7W7D:GZXV:JWJP:I3AF:5VOP:2ET4
Docker Root Dir: /var/lib/docker/165536.165536
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Comment by gambol99** [comment]
Potentially similar issue https://github.com/moby/moby/issues/32820

**Comment by thaJeztah** [comment]
Hm, I see the same error (`Host ID 0 cannot be mapped to a container ID`) also on 17.12.1, but it looks to be resolved on the release candidates for Docker 18.03;

Note that I add a volume for `/var/lib/docker` as well, to prevent running `overlay2` on top of `overlay2` (I'd recommend doing so when running docker-in-docker);

```bash
docker run --privileged -it --rm \
  -v /lib/modules:/lib/modules:ro \
  -v /var/lib/docker \
  docker:18.03-rc-dind dockerd --storage-driver=overlay2 --userns-remap=default
```

Inside the container;

```
docker build --no-cache -t test -<<EOF
FROM alpine
RUN touch /me
RUN rm -fv /me
EOF
```

Then test the result;

```
docker run --rm test ls /me -l
``

Can you try on the 18.03 RC?

**Comment by gambol99** [comment]
Hi @thaJeztah .. confirmed the issue is not present in 18.03-rc-dind and we can live with an RC for now :-) .. Much appreciated!!
